File: mathsys/dev/python/start.py
#^
#^  HEAD
#^

#> HEAD -> MODULES
from dataclasses import dataclass

#> HEAD -> DATA
from .nonterminal import NonTerminal
from .ir import Opcode, Binary, Pointer, Vec, node


#^
#^  START
#^

#> START -> CLASS
@dataclass(init = False, unsafe_hash = True)
class Start(NonTerminal):
    code = Opcode(0x01).binary()
    stream: tuple[Level1, ...]
    def __init__(self, items: list[Level1]) -> None: 
        self.stream = tuple(items)
        self.freeze()
    def latex(self, types: dict[str, str] = {}) -> str:
        match len(self.stream):
            case 0: delimiters = ["", ""]
            case 1: delimiters = [r"\(", r"\)"]
            case other: delimiters = [r"\[", r"\]"]
        values = r"\\ ".join(element.latex(types) for element in self.stream)
        while values.startswith(r"\\"): values = values[2:]
        return f"{delimiters[0]}{values}{delimiters[1]}"
    def ir(self, binary: list[Binary], nodes: list[Binary] = []) -> Pointer:
        pointers = Vec([element.ir(binary, nodes) for element in self.stream])
        return node(self.code + pointers, binary, nodes)

#> START -> TYPES
from .level1 import Level1

File: mathsys/dev/typescript/start.ts
//^
//^ HEAD
//^

//> HEAD -> DATA
import {NonTerminal} from "./nonterminal";
import {Opcode, Binary, Pointer, Vec, node} from "./ir";


//^
//^ START
//^

//> START -> CLASS
export class Start extends NonTerminal {
    private code = new Opcode(0x01).binary();
    readonly stream: Level1[];
    constructor(items: Level1[]) {
        super();
        this.stream = [...items];
        this.freeze();
        return this;
    }
    latex(types: Map<string, string> = new Map<string, string>()): string {
        let delimiters;
        switch (this.stream.length) {
            case 0: {delimiters = ['', '']; break}
            case 1: {delimiters = [String.raw`\(`, String.raw`\)`]; break}
            default: {delimiters = [String.raw`\[`, String.raw`\]`]}
        }
        let values = this.stream.map(element => element.latex(types)).join(String.raw`\\ `);
        while (values.startsWith(String.raw`\\`)) {values = values.slice(2)}
        return `${delimiters[0]}${values}${delimiters[1]}`;
    }
    ir(binary: Binary[], nodes: Binary[] = []): Pointer {
        const pointers = new Vec(this.stream.map(element => element.ir(binary, nodes)));
        return node(this.code.add(pointers), binary, nodes);
    }
}

//> START -> TYPES
import {Level1} from "./level1";